Math Test Suite Part 2

z3dev opened the pull request:
WORK IN PROGRESS

PART 2 of the test suites for math functionality
 - [ ] mat4 IN PROGRESS, 52.61%, ISSUES see comments
 - [X] plane DONE, 86.36 % (no tests for fromData/fromObject/fromPoints, otherwise 100%)
 - [X] vec2 DONE, 90.84 % (no tests for fromVarious, otherwise 100%)
 - [X] vec3 DONE, 89.90 % (no tests for fromVarious), otherwise 100%
 - [X] vec4 DONE, 100.00 % coverage

Need to add tests for issue #17

z3dev commented:
@kaosat-dev please review quickly as there were fixes for vec2. please take a look at each of the tests for the associated fixes. cheers

kaosat-dev commented on src/core/math/vec2/cross.js:
```diff
@@ -24,7 +24,7 @@ function cross (...params) {
   let z = a[0] * b[1] - a[1] * b[0]
   out[0] = out[1] = 0
-  out[2] = z
+  out[1] = z
```
err no a cross product only exists in 3d space, and you want to set the third component (at index 2) to Z , not at index 1

z3dev replied:
Understood, but a vec2 function returning a vec3 result just seems wrong. How about removing this from vec2?

kaosat-dev replied:
no, that is totally normal, that is the maths for cross product, not something we want to change :) there are plenty of functions returning scaler (float) values aswell, it is all fine

kaosat-dev commented on src/core/math/vec2/dot.test.js:
```diff
@@ -0,0 +1,24 @@
+const test = require('ava')
+const {dot, fromValues} = require('./index')
+
+test('vec2: dot() should return correct values', t => {
```
nice and simple !

kaosat-dev commented on src/core/math/vec2/distance.test.js:
```diff
@@ -0,0 +1,30 @@
+const {distance, fromValues} = require('./index')
+
+const {nearlyEqual} = require('../../../../test/helpers/index')
+const {EPS} = require('../../constants')
```
are you sure about the use of EPS here ? we have not looked at that one in  along while

z3dev replied:
I’ve been using EPS for all the nearlyEqual tests, thinking that this would be a better value then just something arbitrary. We can review EPS when working on the geometry / BSP trees, which is where EPS becomes important.

kaosat-dev commented on src/core/math/vec2/cross.test.js:
```diff
@@ -0,0 +1,81 @@
+const test = require('ava')
```
nice all !

kaosat-dev commented on src/core/math/mat4/mirror.js:
```diff
@@ -7,7 +7,7 @@ const fromValues = require('./fromValues')
  * create an affine matrix for mirroring into an arbitrary plane:
```
mirror is complicated, our old & new implementations differ quite a bit

z3dev replied:
I’ve been testing for consistent results as well. If we need to revisit mirror then now is the time. The math functions are the heart and soul of the library.

Do you have an example? I’m sure the API is broken, so we can skip that level.

kaosat-dev replied:
not off the top of my head but it is essential that we keep both implementations in the file for now, until they can be unified/ merged down the line

z3dev commented:
@kaosat-dev please take another look, as I corrected vec2.cross. And added mat4 tests.

THIS NEEDS ATTENTION.. There are some unnecessary functions in mat4.
leftMultiplyVec2.js
leftMultiplyVec3.js
rightMultiplyVec2.js
rightMultiplyVec3.js

mat4.rightMultiply...
V1 uses only helpers.rightMultiply1x3VectorToArray(). If rightMultiply is required then we can add to vec3. So, I think the mat4 functions are never used. Can I drop these in mat4?

mat4.leftMultiply...
V1 Vector2 and Vector3 uses to transform vectors using Matrix4. Why these classes have two functions with the same functionality... I have no idea.

Also, V2 vec2.transformMat4 and vec3.transformMat4 seem to be the new V2 functions. But I think we should review closely. There seem to be slight differences in return values.

z3dev commented:
@kaosat-dev Please review again. I think this PR should be merged as there were a number of changes to bring the functions up to ES6, and JSCAD standards for code.

I will continue on via another PR, as there will be a number of changes in order to make functions consistent across the objects, i.e. further changes to use (...params)

kaosat-dev commented on src/core/math/mat4/clone.js:
```diff
@@ -1,28 +1,18 @@
-  out[15] = a[15]
-  return out
+const clone = (a) => {
+  return fromValues(
```
smaart !

kaosat-dev commented on src/core/math/mat4/create.js:
```diff
@@ -1,12 +1,10 @@
-function create () {
-  let out = new Float32Array(16)
+const create = () => {
+  const out = new Float32Array(16)
```
good use of const !

kaosat-dev commented on src/core/math/mat4/equals.js:
```diff
@@ -5,7 +5,7 @@
  * @returns {Boolean} True if the matrices are equal, false otherwise.
  */
-function exactEquals (a, b) {
+const exactEquals = (a, b) => {
```
yes !

kaosat-dev commented on src/core/math/mat4/equals.test.js:
```diff
@@ -0,0 +1,56 @@
+const test = require('ava')
+const { equals, fromValues } = require('./index')
+
+test('vec3: equals() should return correct booleans', (t) => {
```
correct booleans ?

kaosat-dev commented on src/core/math/mat4/fromRotation.js:
```diff
@@ -38,9 +38,9 @@ function fromRotation (...params) {
-  s = Math.sin(rad)
-  c = Math.cos(rad)
-  t = 1 - c
+  const s = Math.sin(rad)
```
good change

kaosat-dev commented on src/core/math/plane/splitLineBetweenPoints.js:
```diff
@@ -1,17 +1,18 @@
   return result
 }
+
+module.exports = splitLineBetweenPoints
```
good change (in general , regarding all these, good idea to move them after the declaration)
